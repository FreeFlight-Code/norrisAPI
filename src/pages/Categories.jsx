import React from "react";
import {connect} from 'react-redux';

class Categories extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      categories: []
    };
  }

  componentDidMount() {
    fetch("https://api.chucknorris.io/jokes/categories")
      .then(res => res.json())
      .then(categories => this.setState({ categories: categories }));
  }

  categoriesList() {
    const { categories } = this.state;
    return categories.map((category, i) => {
      return (
        <div
          className="category"
          onClick={e => {
            this.getJoke(category);
          }}
          key={`category-${i}`}
        >
          {category}
        </div>
      );
    });
  }

  getJoke(category) {
    if (category) {
      fetch(`https://api.chucknorris.io/jokes/random?category=${category}`)
        .then(res => res.json())
        .then(joke => {
          this.props.dispatch({
            type: "DISPLAY_MODAL",
            message: joke.value,
            messageType: "info"
          });
        });
    }
  }

  render() {
    return (
      <div className="page categories">
        <div className="categoryList">{this.categoriesList()}</div>
        {this.getJoke()}
      </div>
    );
  }
}

export default connect()(Categories);
