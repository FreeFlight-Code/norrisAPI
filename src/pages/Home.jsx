import React from 'react';
import Form from '../components/Form';

class Home extends React.Component {

	render() { 

		console.log(this.state)
		return (

      <div className="page home">
        Home Page
		<Form/>
      </div>
    );
	}
}
 
export default Home;