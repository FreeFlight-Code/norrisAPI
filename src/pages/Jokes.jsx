
import React, { Component } from "react";

class Jokes extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }
  render() {
    return <div>Jokes Page</div>;
  }
}

export default Jokes;
