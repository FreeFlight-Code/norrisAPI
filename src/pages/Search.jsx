import React, { Component } from "react";
import {connect} from 'react-redux';

class Search extends Component {
  constructor(props) {
    super(props);
    this.state = {
      searchTerm: "",
      jokes: []
    };
  }

  handleinputChange(val){
    this.setState({
      searchTerm: val
    })
  }

  handleSubmit(){
    const {searchTerm} = this.state;
    fetch(`https://api.chucknorris.io/jokes/search?query=${searchTerm}`)
      .then(res => res.json())
      .then(obj => {
        let jokesArray = obj.result;
        this.setState({ 
          searchTerm: "",
          jokes: jokesArray
        })
      });
  }

  truncJoke(string, charLimit){
    return string.slice(0, charLimit + 1) + "..."
  }

  renderResults(){
    const {jokes} = this.state;
    if(jokes && jokes.length > 0){
      return jokes.map((joke, i)=>{
        let shortJoke = this.truncJoke(joke.value, 50);
        return (
          <div
            onClick={e => this.props.dispatch({ type: "DISPLAY_MODAL", message: joke.value, messageType: "info" })}
            key={`joke-${i}`}
          >
            {shortJoke}
          </div>
        );
      })
    }
  }

  render() {
    return (
      <div className="page search">
        <label htmlFor="searchBar">Search Bar</label>
        <input value={this.state.searchTerm} id="searchBar" type="text" onChange={e=>{this.handleinputChange(e.target.value)}} placeholder="Search Chuck Norris Joke by Key Word"></input>
        <button className="blue" onClick={e=>{this.handleSubmit()}}>Submit</button>
        <div className="results">
          <h3>Results</h3>
          {this.renderResults()}
        </div>
      </div>
    )
  }
}

export default connect()(Search);