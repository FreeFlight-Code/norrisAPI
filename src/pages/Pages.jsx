import React from "react";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import { connect } from "react-redux";
import Home from "./Home";
import Categories from "./Categories";
import Jokes from "./Jokes";
import Search from "./Search";

class Routes extends React.Component {

  render() {
    const {loggedin} = this.props.user;
    return (
      <Router>
        {loggedin && <Switch>
          <Route component={Categories} path="/categories" />
          <Route component={Jokes} path="/jokes" />
          <Route component={Search} path="/search" />
        </Switch>}
        <Switch>
          <Route component={Home} exact path="/" />
          <Route path="/">
            <div>404 sorry page not found</div>
          </Route>
        </Switch>
      </Router>
    );
  }
}
 

const mapStateToProps = state => {
  return {
    user: state.user
  };
};

export default connect(mapStateToProps)(Routes);