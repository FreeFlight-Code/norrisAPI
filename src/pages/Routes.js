import React from "react";
import {
  Route,
  Switch,
  useLocation
} from "react-router-dom";
import { connect } from "react-redux";
//components
import Categories from './Categories';
import Jokes from './Jokes';
import Search from './Search';
import Home from './Home';

//#functional component
function Routes(props) {
  //#destructuring
  const {loggedin} = props;
  return (
    <>
{ loggedin && //secured paths only
      <Switch>
        <Route exact path="/categories">
          <Categories />
        </Route>
        <Route exact path="/jokes">
          <Jokes />
        </Route>
        <Route exact path="*">
          <Search />
        </Route>
      </Switch>}

{!loggedin && // non-secured paths only
      <Switch> 
        <Route exact path="/">
          <Home />
        </Route>
        <Route path="*">
          <NoMatch />
        </Route>
      </Switch>}
    </>
  );
}

function NoMatch() {
  let location = useLocation();

  return (
    <div className="page">
      <h3>
        No match for <code>{location.pathname}</code>
      </h3>
      <h3>
        Please login to access secure routes or check your url.
      </h3>
    </div>
  );
}

const mapStateToProps = state =>({
  loggedin: state.user.loggedin
})

export default connect(mapStateToProps)(Routes)