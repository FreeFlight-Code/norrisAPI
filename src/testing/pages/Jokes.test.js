import React from "react";
import {JokesPage, List, Joke} from "../../pages/Jokes";
import renderer from "react-test-renderer";
import { cleanup } from "@testing-library/react";

jest.mock("../../components/LoginForm.jsx");

var history = []
beforeEach(()=>{
	history = [
    {
      categories: ["turtle"],
      viewed_at: new Date(1997, 1, 2, 1, 30),
      value: "what up",
    },
    {
      categories: ["apple"],
      viewed_at: new Date(1998, 1, 2, 14, 30),
      value: "always smile",
    },
    {
      categories: ["water"],
      viewed_at: new Date(1999, 1, 2, 10, 30),
      value: "COVID-19",
    },
    {
      categories: ["giraffe"],
      viewed_at: new Date(1990, 1, 2, 4, 30),
      value: "peace on earth",
    },
    {
      categories: ["bear"],
      viewed_at: new Date(1994, 1, 2, 2, 30),
      value: "whello world",
    }
  ];
  cleanup();
})

describe(" The Jokes Page... ", () => {
  it("page renders correctly", () => {
    const tree = renderer.create(<JokesPage />).toJSON();
    expect(tree).toMatchSnapshot();
    const ren = renderer.create(<JokesPage />);
  });
  it("list renders correctly", () => {
    const tree = renderer.create(<List history={history}/>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("joke renders correctly", () => {
    const tree = renderer.create(<Joke {...history[0]}/>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("renders all items in history", ()=>{
    const tree = renderer.create(<List history={history} />).toJSON();
    expect(tree.length).toEqual(5)
  })
  it("renders first category if more than one", ()=>{
    const joke = renderer
      .create(
        <Joke
              categories={["giraffe", "red", "jetBlue"]}
              value={"peace on earth"}
              viewed_at={new Date(1990, 1, 2, 4, 30)}
        />
      );
      const root = joke.root;
    const className = root.findByProps({className: "category"});
    expect(className.props.children).toEqual("giraffe")
  })
  it("expect if no date provided to set date to \"Invalid Date\"", ()=>{
    const joke = renderer
      .create(
        <Joke
              categories={["giraffe"]}
              value={"peace on earth"}
              viewed_at={""}
        />
      );
      const root = joke.root;
    const viewed_at = root.findByProps({className: "viewed_at"});
    expect(viewed_at.props.children).toEqual("Invalid Date")
  })
});
