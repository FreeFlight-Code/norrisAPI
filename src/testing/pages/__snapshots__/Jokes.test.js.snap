// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` The Jokes Page...  joke renders correctly 1`] = `
<div
  className="joke"
>
  <span>
    what up
  </span>
  <span
    className="category"
  >
    turtle
  </span>
  <span
    className="viewed_at"
  >
    1:30:00 AM
  </span>
</div>
`;

exports[` The Jokes Page...  list renders correctly 1`] = `
Array [
  <div
    className="joke"
  >
    <span>
      what up
    </span>
    <span
      className="category"
    >
      turtle
    </span>
    <span
      className="viewed_at"
    >
      1:30:00 AM
    </span>
  </div>,
  <div
    className="joke"
  >
    <span>
      always smile
    </span>
    <span
      className="category"
    >
      apple
    </span>
    <span
      className="viewed_at"
    >
      2:30:00 PM
    </span>
  </div>,
  <div
    className="joke"
  >
    <span>
      COVID-19
    </span>
    <span
      className="category"
    >
      water
    </span>
    <span
      className="viewed_at"
    >
      10:30:00 AM
    </span>
  </div>,
  <div
    className="joke"
  >
    <span>
      peace on earth
    </span>
    <span
      className="category"
    >
      giraffe
    </span>
    <span
      className="viewed_at"
    >
      4:30:00 AM
    </span>
  </div>,
  <div
    className="joke"
  >
    <span>
      whello world
    </span>
    <span
      className="category"
    >
      bear
    </span>
    <span
      className="viewed_at"
    >
      2:30:00 AM
    </span>
  </div>,
]
`;

exports[` The Jokes Page...  page renders correctly 1`] = `
<div
  className="page jokes"
>
  <div
    className="filters"
  >
    <span
      onClick={[Function]}
    >
      By Category
    </span>
    <span
      onClick={[Function]}
    >
      Most Recent
    </span>
    <span
      onClick={[Function]}
    >
      First Viewed
    </span>
  </div>
  <h3>
    All jokes viewed
  </h3>
  <div
    className="joke"
  >
    <span>
      Joke
    </span>
    <span
      className="category"
    >
      Category
    </span>
    <span>
      Time
    </span>
  </div>
</div>
`;
